src/Sarif/HashUtilities.cs

```diff
@@ -2,21 +2,71 @@
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
 using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
 using System.Diagnostics.CodeAnalysis;
 using System.IO;
 using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
 
 namespace Microsoft.CodeAnalysis.Sarif
 {
     public static class HashUtilities
     {
+        static HashUtilities() => FileSystem = new FileSystem();
+
+        private static IFileSystem _fileSystem;
+        internal static IFileSystem FileSystem
+        {
+            get
+            {
+                _fileSystem = _fileSystem ?? new FileSystem();
+                return _fileSystem;
+            }
+
+            set
+            {
+                _fileSystem = value;
+            }
+        }
```

Comment:
I don't think you need all of this. Since the static ctor initializes the property, it's guaranteed to be set by the time anyone invokes the getter, so
```
internal static IFileSystem FileSystem { get; set; }
```
will suffice.

Reply:
this doesn't provide for the pattern where a user wants to remove their file system mock, so we set the property to null. this allows tests to set the mock, then set to null, after which the code will naturally rehydrate the standard file system object.

---
In reply to: [303174647](https://github.com/microsoft/sarif-sdk/pull/1574#discussion_r303174647) [](ancestors = 303174647)

Reply:
That is compelling and requires what you just said to be a code comment :-)

---
In reply to: [303175028](https://github.com/microsoft/sarif-sdk/pull/1574#discussion_r303175028) [](ancestors = 303175028,303174647)

Reply:
Reactivating this for your further consideration. At the very least, if you're not persuaded by this argument, please add the code comment to explain the scenario you're enabling.

I don't think the test code pattern you're enabling is a good one:
```
HashUtilities.FileSystem = mockFileSystem;

// Do stuff

HashUtilities.FileSystem = null;

// Good news! The standard file system will return on the next get access to this property. But I doubt the client knows that.
```

I think that behavior violates least surprise. It could even cause a bug if (because this is test code, after all) the value of this property was _not_ a standard `FileSystem` object before you got your hands on it. I prefer the "leave things as you found them" pattern:
```
IFileSystem previousFileSystem = HashUtilitles.FileSystem;
HashUtilities.FileSystem = mockFileSystem;

// Do stuff

HashUtilities.FileSystem = previousFileSystem;

// Good news! You know for a fact you've done no harm.
```

---
In reply to: [303175558](https://github.com/microsoft/sarif-sdk/pull/1574#discussion_r303175558) [](ancestors = 303175558,303175028,303174647)